memory_allocator.c

```diff
@@ -0,0 +1,203 @@
+}
+
+void malloc_add_arena(malloc_pool_t *mp, void *start, size_t arena_size) {
+  if (arena_size < sizeof(mem_arena_t))
+    return;
+
+  memset(start, 0, sizeof(mem_arena_t));
+  mem_arena_t *ma = start;
+
+  TAILQ_INSERT_HEAD(&mp->mp_arena, ma, ma_list);
+  ma->ma_size = arena_size - sizeof(mem_arena_t);
+
+  TAILQ_INIT(&ma->ma_freeblks);
+  TAILQ_INIT(&ma->ma_usedblks);
+
+  // Adding the first free block that covers all the remaining arena_size.
+  mem_block_t *mb = (mem_block_t *)((char *)ma + sizeof(mem_arena_t));
+  size_t block_size = arena_size - sizeof(mem_arena_t);
+  add_free_memory_block(ma, mb, block_size);
+}
+
+static mem_block_t *find_entry(struct mb_list *mb_list, size_t total_size) {
+  mem_block_t *current = NULL;
+  TAILQ_FOREACH(current, mb_list, mb_list) {
+    if (current->mb_size >= total_size)
+      return current;
+  }
+  return NULL;
+}
+
+static mem_block_t *try_allocating_in_area(mem_arena_t *ma,
+    size_t requested_size, uint16_t flags) {
+  mem_block_t *mb = find_entry(&ma->ma_freeblks,
+                               requested_size + sizeof(mem_block_t));
+
+  if (!mb) /* No entry has enough space. */
+    return NULL;
+
+  TAILQ_REMOVE(&ma->ma_freeblks, mb, mb_list);
+  size_t total_size_left = mb->mb_size - requested_size;
+  if (total_size_left > sizeof(mem_block_t)) {
+    mb->mb_size = requested_size;
+    mem_block_t *new_mb = (mem_block_t *)((char *)mb + requested_size + sizeof(
+                                            mem_block_t));
+    add_free_memory_block(ma, new_mb, total_size_left);
+  }
+
+  return mb;
+}
+
+void *allocate(size_t size, malloc_pool_t *mp, uint16_t flags) {
+  size_t size_aligned;
+  if (size % ALIGNMENT == 0)
+    size_aligned = size;
+  else
+    size_aligned = size + ALIGNMENT - size % ALIGNMENT;
+
+  /* Search for the first area in the list that has enough space. */
+  mem_arena_t *current = NULL;
+  TAILQ_FOREACH(current, &mp->mp_arena, ma_list) {
+    mem_block_t *mb = try_allocating_in_area(current, size_aligned, flags);
+    if (mb) {
+      add_used_memory_block(current, mb);
+
+      if (flags == M_ZERO)
+        memset(mb->mb_data, 0, size);
+
+      return mb->mb_data;
+    }
+  }
+
+  return NULL;
+}
+
+void deallocate(void *addr, malloc_pool_t *mp) {
+  mem_block_t *mb = (mem_block_t *)(((char *)addr) - sizeof(mem_block_t));
+
+  if (mb->mb_magic != MAGIC || mp->mp_magic != MAGIC)
+    kprintf("Memory corruption detected!\n");
+
+  mem_arena_t *current = NULL;
+  TAILQ_FOREACH(current, &mp->mp_arena, ma_list) {
+    char *start = ((char *)current) + sizeof(mem_arena_t);
+    if ((char *)addr >= start && (char *)addr < start + current->ma_size) {
+      remove_used_memory_block(current, mb);
+      add_free_memory_block(current, mb, mb->mb_size + sizeof(mem_block_t));
+    }
+  }
+}
+
+void print_free_blocks(malloc_pool_t *mp) {
+  mem_arena_t *current_arena = NULL;
+  TAILQ_FOREACH(current_arena, &mp->mp_arena, ma_list) {
```

Comment:
I haven't implemented the "size is less than zero, so it's a free block" feature yet, because I am not convinced that it's better than simply another field boolean field in the struct. In the current implementation, the size is always > 0.
Should I implement it?